modules/mod_seo_sitemap/models/m_seo_sitemap.erl

```diff
@@ -0,0 +1,478 @@
+%% @author Marc Worrell <marc@worrell.nl>
+%% @copyright 2021 Marc Worrell
+%% @doc Model for tracking all URLs to be included in the sitemap.
+
+%% Copyright 2021 Marc Worrell
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(m_seo_sitemap).
+
+-export([
+    m_find_value/3
+    ]).
+
+-export([
+    count/1,
+    max_id/1,
+    slice/3,
+    update/2,
+    delete_key/3,
+    delete_loc/2,
+
+    update_rsc/2,
+    rebuild_rsc/1,
+    install/1
+    ]).
+
+-export([
+    rebuild_rsc_task/2
+    ]).
+
+-include_lib("zotonic.hrl").
+-include("../include/seo_sitemap.hrl").
+
+m_find_value(urlsets, #m{ value = undefined }, Context) ->
+    seo_sitemap:urlsets(Context);
+m_find_value(urlset, #m{ value = undefined } = M, _Context) ->
+    M#m{ value = urlset };
+m_find_value(Source, #m{ value = urlset } = M, _Context) ->
+    M#m{ value = {urlset, z_convert:to_binary(Source)} };
+m_find_value(Offset, #m{ value = {urlset, Source} } = M, _Context) ->
+    Offs = case Offset of
+        undefined -> 1;
+        _ ->
+            try z_convert:to_integer(Offset)
+            catch _:_ -> 1
+            end
+    end,
+    Offs1 = erlang:max(1, Offs),
+    M#m{ value = {urlset, Source, Offs1} };
+m_find_value(Limit, #m{ value = {urlset, Source, Offset} }, Context) ->
+    Limit1 = case Limit of
+        undefined -> ?URLSET_SIZE;
+        _ ->
+            try z_convert:to_integer(Limit)
+            catch _:_ -> ?URLSET_SIZE
+            end
+    end,
+    Limit2 = erlang:min(?URLSET_SIZE, Limit1),
+    seo_sitemap:urlset(Source, Offset, Limit2, Context).
+
+%% @doc Return the number of sitemap entries
+-spec count( z:context() ) -> non_neg_integer().
+count( Context ) ->
+    z_db:q1("select count(*) from seo_sitemap", Context).
+
+-spec max_id( z:context() ) -> non_neg_integer().
+max_id( Context ) ->
+    z_db:q1("select max(id) from seo_sitemap", Context).
+
+
+%% @doc Return a slice from the sitemap entries. Note that this might
+%% give duplicate entries in sitemaps, as the ids might change during
+%% sitemap discovery and consumption by the crawler.
+%% All locations are made into absolute urls with the correct hostname.
+%% As entries are filtered on publication_end, their language and resource
+%% visibility, this might return less entries than the Limit amount requested.
```

Comment:
Agreed that this is a workable situation. To easily prevent concurrency issues, we could use a snapshot view on the sitemap data, which would be natural practice in a versioned database. For the current implementation, the omission is ok to keep it simple.